util/stmtsummary/evicted.go

```diff
@@ -0,0 +1,318 @@
+	ssElement := new(stmtSummaryByDigestElement)
+	ssElement.beginTime = beginTimeForCurrentInterval
+	ssElement.endTime = beginTimeForCurrentInterval + intervalSeconds
+	return &stmtSummaryByDigestEvictedElement{
+		digestKeyMap: make(map[string]struct{}),
+		sum:          ssElement,
+	}
+}
+
+// AddEvicted is used add an evicted record to stmtSummaryByDigestEvicted
+func (ssbde *stmtSummaryByDigestEvicted) AddEvicted(evictedKey *stmtSummaryByDigestKey, evictedValue *stmtSummaryByDigest, historySize int) {
+
+	// *need to get optimized*!!
+	evictedValue.Lock()
+	defer evictedValue.Unlock()
+	for e := evictedValue.history.Back(); e != nil; e = e.Prev() {
+		eBeginTime := e.Value.(*stmtSummaryByDigestElement).beginTime
+		eEndTime := e.Value.(*stmtSummaryByDigestElement).endTime
+
+		// prevent exceeding history size
+		for ssbde.history.Len() >= historySize && ssbde.history.Len() > 1 {
+			ssbde.history.Remove(ssbde.history.Front())
+		}
+
+		// look for matching history interval
+		if ssbde.history.Len() == 0 && historySize > 0 {
+			// no record in history
+			beginTime := eBeginTime
+			intervalSeconds := eEndTime - eBeginTime
+			record := newStmtSummaryByDigestEvictedElement(beginTime, intervalSeconds)
+			record.addEvicted(evictedKey, e.Value.(*stmtSummaryByDigestElement))
+			ssbde.history.PushBack(record)
+
+			if evictedKey == nil {
+				// passing a empty Key is used as `refresh`.
+				continue
+			}
+		}
+
+		for h := ssbde.history.Back(); h != nil; h = h.Prev() {
+			sBeginTime := h.Value.(*stmtSummaryByDigestEvictedElement).sum.beginTime
+			sEndTime := h.Value.(*stmtSummaryByDigestEvictedElement).sum.endTime
+
+			if sBeginTime <= eBeginTime &&
+				sEndTime >= eEndTime {
+				// is in this history interval
+				h.Value.(*stmtSummaryByDigestEvictedElement).addEvicted(evictedKey, e.Value.(*stmtSummaryByDigestElement))
+				break
+			}
+
+			if sEndTime <= eBeginTime {
+				// digest is young, insert into new interval after this history interval
+				beginTime := eBeginTime
+				intervalSeconds := eEndTime - eBeginTime
+				record := newStmtSummaryByDigestEvictedElement(beginTime, intervalSeconds)
+				record.addEvicted(evictedKey, e.Value.(*stmtSummaryByDigestElement))
+				ssbde.history.InsertAfter(record, h)
+				break
+			}
+
+			if sBeginTime > eEndTime {
+				// digestElement is old
+				if h != ssbde.history.Front() {
+					// check older history digestEvictedElement
+					continue
+				} else if ssbde.history.Len() >= historySize {
+					// out of history size, abandon
+					break
+				} else {
+					// is oldest digest
+					// creat a digestEvictedElement and PushFront!
+					beginTime := eBeginTime
+					intervalSeconds := eEndTime - eBeginTime
+					record := newStmtSummaryByDigestEvictedElement(beginTime, intervalSeconds)
+					record.addEvicted(evictedKey, e.Value.(*stmtSummaryByDigestElement))
+					ssbde.history.PushFront(record)
+					break
+				}
+			}
```

Comment:
Does any test cover this? Because I think this situation should never happen.

Reply:
currently no.